### PR TITLE
[WIP] [Fix] copy does not check users reference

### PR DESCRIPTION
**Pull request by jonasheinrich**
Fixes https://www.openproject.org/work_packages/1653

### CODE REVIEW

**Comment by mugwump** [comment]
&nbsp;<b>Code Climate</b> has <a href="https://codeclimate.com/repos/528f35c8c7f3a33b7d03cde8/compare/fix/copy_does_not_check_users_reference">analyzed this pull request</a>.

File: app/models/work_package/validations.rb [hunk 1]
```diff
@@ -101,6 +104,22 @@ def validate_active_priority
       errors.add :priority_id, :only_active_priorities_allowed
     end
   end
+  
+  def validate_assigned_to_is_member
+    if self.assigned_to.present?
+      unless self.project.principals.map(&:id).include?(self.assigned_to.id)
```
Comment (ulferts) [inline app/models/work_package/validations.rb]: There is the `assignable_assignees`-method on WorkPackage that is more precise than your `project.principals`.

Additionally one could refactor this to create an `assignable_to?(user)`-method, that would capsulate what you are trying to do here.

One thing that we will have to talk about is the consequence of adding such a validation to an existing data set. There will be a lot of invalid work packages once this is activated.

File: app/models/work_package/validations.rb [hunk 2]
```diff
@@ -101,6 +104,22 @@ def validate_active_priority
...
+        errors.add(:assigned_to, l(:error_must_be_project_member))
+      end
+    end
+  end
+  
+  def validate_responsible_is_member
+    unless self.responsible.nil?
+      unless self.project.principals.map(&:id).include?(self.responsible.id)
```
Comment (ulferts) [inline app/models/work_package/validations.rb]: See above

**Comment by jonasheinrich** [comment]
When we merge this a lot of work packages will become invalid. I think this is no option. 
There are at least two ways of avoiding this: Deleting all assignee/responsible relations to users that are not in the project or adding all those users to the project. 
Maybe it even depends on the previously installed plugins (e.g. Impermanent Membership). 
One solution could be: 
- Change the possible assignee/responsible in that way that a user U is a possible assignee/responsible for a work package W, if U is a member of the project of W or this project is public. 
- This way only work packages in private projects can get invalid. 
- But if W is a private project U did not know anything about W. So the assignee/responsible relation was ineffective before. 
- This way one could delete the respective assignee/responsible relations in private projects.
- Adding users to a project seems way worse. But maybe there is another solution.

File: spec/models/work_package/work_package_copy_spec.rb [hunk 3]
```diff
@@ -179,6 +179,21 @@
...
     end
+    
+    describe "a wp with assignee X to a project X is not member of" do
+      let(:source_project) { FactoryGirl.create(:project) }
+      let(:target_project) { FactoryGirl.create(:project) }
+      let(:assignee) { FactoryGirl.create(:user, 
+      									  :member_in_project => source_project) }
+      let(:work_package) { FactoryGirl.build(:work_package, 
+      										  :project => source_project,
+      										  :assigned_to => assignee) }
```
Comment (myabc) [inline spec/models/work_package/work_package_copy_spec.rb]: @jonasheinrich it'd be better to stick to spaces here.

**Comment by myabc** [comment]
@jonasheinrich is this PR still relevant?

**Pull request closed without merging**